cpp/src/io/parquet/reader_impl_preprocess.cu

```diff
@@ -307,6 +307,90 @@ void generate_depth_remappings(std::map<int, std::pair<std::vector<int>, std::ve
   return total_pages;
 }
 
+/**
+ * @brief Count the total number of pages using page index information.
+ */
+[[nodiscard]] size_t count_page_headers_with_pgidx(host_span<ColumnChunkDesc> chunks)
+{
+  size_t total_pages = 0;
+  for (auto& chunk : chunks) {
+    CUDF_EXPECTS(chunk.h_chunk_info != nullptr, "Expected non-null column info struct");
+    auto const& chunk_info = *chunk.h_chunk_info;
+    chunk.num_dict_pages   = chunk_info.has_dictionary() ? 1 : 0;
+    chunk.num_data_pages   = chunk_info.pages.size();
+    total_pages += chunk.num_data_pages + chunk.num_dict_pages;
+  }
+  return total_pages;
+}
+
+// struct used to carry info from the page indexes to the device
+struct page_index_info {
+  int32_t num_rows;
+  int32_t chunk_row;
+  int32_t num_nulls;
+  int32_t num_valids;
+  int32_t str_bytes;
+};
+
+// functor to copy page_index_info into the PageInfo struct
+struct copy_page_info {
+  device_span<page_index_info> page_info;
+  device_span<PageInfo> pages;
+
+  __device__ void operator()(size_type idx)
+  {
+    auto& pg                = pages[idx];
+    auto const& pi          = page_info[idx];
+    pg.num_rows             = pi.num_rows;
+    pg.chunk_row            = pi.chunk_row;
+    pg.has_page_index       = true;
+    pg.num_nulls            = pi.num_nulls;
+    pg.num_valids           = pi.num_valids;
+    pg.str_bytes_from_index = pi.str_bytes;
+    pg.str_bytes            = pi.str_bytes;
+    pg.start_val            = 0;
+    pg.end_val              = pg.num_valids;
+  }
+};
+
+/**
+ * @brief Set fields on the pages that can be derived from page indexes.
+ *
+ * This replaces some preprocessing steps, such as page string size calculation.
+ */
+void fill_in_page_info(host_span<ColumnChunkDesc> chunks,
+                       device_span<PageInfo> pages,
+                       rmm::cuda_stream_view stream)
+{
+  auto const num_pages = pages.size();
+  std::vector<page_index_info> page_info(num_pages);
```

Comment:
oh no, the `page_index_info page_info` and `PageInfo page` conundrum applies here as well. Same as above, if pages are dropping their info part of the type name, maybe page_info should as well and be page_index :shrug: